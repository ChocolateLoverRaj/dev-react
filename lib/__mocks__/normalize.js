throw new Error('normalize is not setup yet')
export default 3