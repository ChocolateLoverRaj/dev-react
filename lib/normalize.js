// Like path.normalize, but removes trailing slashes and backslashes
import { normalize } from 'path'

const trailingSlashRegex = /[\\\/]$/

export default s => normalize(s).replace(trailingSlashRegex, '')
