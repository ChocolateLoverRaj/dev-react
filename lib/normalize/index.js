// Like path.normalize, but removes trailing (slashes and backslashes)
// Also converts all backslashes to slashes
const backslashRegex = /\\/g
const multipleSlashRegex = /\/+/g
const dotSlashRegex = /\/\.\//g
const dotDotSlashRegex = /[^/]*\/\.{2}\//g
const trailingSlashRegex = /\/$/

export default p => p
  .replace(backslashRegex, '/')
  .replace(multipleSlashRegex, '/')
  .replace(dotSlashRegex, '/')
  .replace(dotDotSlashRegex, '')
  .replace(trailingSlashRegex, '')
