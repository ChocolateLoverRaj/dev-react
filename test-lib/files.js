export class File {
  constructor () {
    this.canRead = true
    this.canWrite = true
  }
}

export class NormalFile extends File {
  constructor (content) {
    super()
    this.content = content
  }
}

export class Dir extends File {
  constructor () {
    super()
    this.files = new Map()
  }
}

class EnoentError extends Error {
  constructor () {
    super(...arguments)
    this.code = 'ENOENT'
  }
}

export const topDir = new Dir()

export const reset = () => {
  topDir.files.clear()
}

export const getFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
  return file
}

export const setFile = (path, fileToSet) => {
  let dir = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (fileTree.length === filenames.length) {
      dir.files.set(filename, fileToSet)
      return
    }
    if (dir.files.has(filename)) {
      dir = dir.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}

export const unlinkFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      if (fileTree.length === filenames.length) {
        file.files.delete(filename)
        return
      }
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}
