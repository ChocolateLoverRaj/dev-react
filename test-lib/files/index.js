// TODO: structure with folders and files
const topDir = new Map()

export default topDir