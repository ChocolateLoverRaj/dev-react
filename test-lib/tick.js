// Returns a promise that gets resolved after process.nextTick()

const tick = () => new Promise(resolve => {
  process.nextTick(resolve)
})

module.exports = tick
