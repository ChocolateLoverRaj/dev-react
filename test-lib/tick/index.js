// Returns a promise that gets resolved after process.nextTick()

const tick = () => new Promise(resolve => {
  process.nextTick(resolve)
})

export default tick
