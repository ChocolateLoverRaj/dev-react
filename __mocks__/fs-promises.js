// Mocker for fs/promises
import { NormalFile, reset, unlinkFile, setFile, getFile } from '../test-lib/files.js'
import { constants } from './fs.js'
import EventEmitter from 'eventemitter3'

export const _frozen = new Set()
export const _errorFiles = new Set()
export const _reset = () => {
  _frozen.clear()
  _errorFiles.clear()
  reset()
}

export const _mock = new EventEmitter()
  .on('unfreeze', filename => {
    _frozen.delete(filename)
  })

const onceUnfrozen = filename => new Promise(resolve => {
  const handler = f => {
    if (f === filename) {
      _mock.off('unfreeze', handler)
      resolve()
    }
  }
  _mock.on('unfreeze', handler)
})

export const access = async (filename, permissions) => {
  const file = getFile(filename)
  const filePermissions = file.canRead * constants.R_OK | file.canWrite * constants.W_OK
  if (permissions !== permissions & filePermissions) {
    throw new Error('Bad permissions.')
  }
}

export const writeFile = async (filename, content) => {
  if (_errorFiles.has(filename)) {
    throw new Error('Error writing file.')
  }
  const write = () => {
    setFile(filename, new NormalFile(content))
    _mock.emit('wrote', filename, content)
  }
  if (_frozen.has(filename)) {
    await onceUnfrozen(filename)
    write()
  } else {
    write()
  }
}

export const unlink = async filename => {
  if (_errorFiles.has(filename)) {
    throw new Error('Error unlinking file.')
  }
  unlinkFile(filename)
}
