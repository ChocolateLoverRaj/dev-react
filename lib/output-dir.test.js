/* global afterEach, test, expect, describe */

import OutputDir from './output-dir.js'
import normalize from './normalize.js'
import { setFile, Dir, NormalFile, reset, getFile } from '../test-lib/files.js'
import tick from '../test-lib/tick.js'
import noResolve from 'no-resolve'
import readDir, { _reset } from './__mocks__/read-create-dir.js'
import fsExtra, { _reset as resetFsExtra } from '../__mocks__/fs-extra.js'
import { mkdir, rmdir, _reset as resetFsPromises } from '../__mocks__/fs-promises.js'

afterEach(() => {
  reset()
  _reset()
  resetFsExtra()
  resetFsPromises()
})

test('initial task', () => {
  const outputDir = new OutputDir('dev')
  expect(outputDir.task.getLines()).toMatchSnapshot()
})

test('empty', async () => {
  setFile('dev', new Dir())
  setFile('dev/dir', new Dir())
  setFile('dev/file', new NormalFile())
  const outputDir = new OutputDir('dev')
  let updateLines
  outputDir.task.emitter.on('update', () => {
    updateLines = outputDir.task.getLines()
  })
  outputDir.empty()
  expect(readDir.calledOnceWith('dev')).toBe(true)
  await outputDir.read
  const readLines = outputDir.task.getLines()
  expect(updateLines).toStrictEqual(readLines)
  expect(readLines).toMatchSnapshot()
  expect(fsExtra.remove.calledTwice).toBe(true)
  expect(normalize(fsExtra.remove.firstCall.firstArg)).toBe('dev/dir')
  expect(normalize(fsExtra.remove.secondCall.firstArg)).toBe('dev/file')
  await outputDir.deletePromise
  const deleteLines = outputDir.task.getLines()
  expect(updateLines).toStrictEqual(deleteLines)
  expect(deleteLines).toMatchSnapshot()
  expect(getFile('dev').files.size).toBe(0)
})

describe('createDir', () => {
  describe('before read', () => {
    test('no delete', async () => {
      setFile('dev', new Dir())
      setFile('dev/page', new Dir())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      await outputDir.createDir('page')
      expect(fsExtra.remove.notCalled).toBe(true)
    })

    test('file to dir', async () => {
      setFile('dev', new Dir())
      setFile('dev/page', new NormalFile())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      await outputDir.createDir('page')
      expect(fsExtra.remove.calledOnce).toBe(true)
      expect(normalize(fsExtra.remove.firstCall.firstArg)).toBe('dev/page')
      expect(mkdir.calledOnce).toBe(true)
      expect(normalize(mkdir.firstCall.firstArg)).toBe('dev/page')
    })

    test('create dir', async () => {
      setFile('dev', new Dir())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      await outputDir.createDir('page')
      expect(mkdir.calledOnce).toBe(true)
      expect(normalize(mkdir.firstCall.firstArg)).toBe('dev/page')
    })
  })

  describe('after read', () => {
    test('deletes, then creates', async () => {
      setFile('dev', new Dir())
      setFile('dev/page', new Dir())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      await outputDir.read
      await outputDir.createDir('page')
      expect(fsExtra.remove.calledOnce).toBe(true)
      expect(mkdir.calledAfter(fsExtra.remove)).toBe(true)
      expect(mkdir.calledOnce).toBe(true)
      expect(normalize(mkdir.firstCall.firstArg)).toBe('dev/page')
    })

    test('creates', async () => {
      setFile('dev', new Dir())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      await outputDir.read
      await outputDir.createDir('page')
      expect(mkdir.calledOnce).toBe(true)
      expect(normalize(mkdir.firstCall.firstArg)).toBe('dev/page')
    })
  })
})

describe('removeDir', () => {
  describe('before read', () => {
    test('after creating', async () => {
      setFile('dev', new Dir())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      outputDir.createDir('page')
      await outputDir.removeDir('page')
      expect(fsExtra.remove.notCalled).toBe(true)
      expect(mkdir.notCalled).toBe(true)
    })

    test('without creating', async () => {
      setFile('dev', new Dir())
      setFile('dev/page', new Dir())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      await outputDir.removeDir('page')
      expect(fsExtra.remove.calledOnce).toBe(true)
      expect(normalize(fsExtra.remove.firstCall.firstArg)).toBe('dev/page')
    })
  })

  describe('after read', () => {
    test('already deleting', async () => {
      setFile('dev', new Dir())
      setFile('dev/page', new Dir())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      await outputDir.read
      await outputDir.removeDir('page')
      expect(fsExtra.remove.calledOnce).toBe(true)
    })

    test('delete', async () => {
      setFile('dev', new Dir())
      const outputDir = new OutputDir('dev')
      outputDir.empty()
      await outputDir.read
      await outputDir.createDir('page')
      await outputDir.removeDir('page')
      expect(fsExtra.remove.notCalled).toBe(true)
      expect(rmdir.calledOnce).toBe(true)
      expect(normalize(rmdir.firstCall.firstArg)).toBe('dev/page')
    })
  })
})

test('destroy', async () => {
  const outputDir = new OutputDir('dev')
  let linesUpdate
  outputDir.task.emitter.on('update', () => {
    linesUpdate = outputDir.task.getLines()
  })
  outputDir.empty()
  await outputDir.destroy()
  await noResolve(tick(), Promise.all([outputDir.read, outputDir.deletePromise]))
  const lines = outputDir.task.getLines()
  expect(linesUpdate).toStrictEqual(lines)
  expect(lines).toMatchSnapshot()
})
