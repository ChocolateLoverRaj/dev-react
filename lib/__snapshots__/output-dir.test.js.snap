// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`destroy 1`] = `Array []`;

exports[`empty 1`] = `
Array [
  "[93mPrepare output dir[39m",
  "[32m   Read output dir[39m",
  "[93m   Delete unnecessary files[39m",
]
`;

exports[`empty 2`] = `
Array [
  "[32mPrepare output dir[39m",
  "[32m   Read output dir[39m",
  "[32m   Delete unnecessary files[39m",
]
`;

exports[`initial task 1`] = `
Array [
  "[93mPrepare output dir[39m",
  "[93m   Read output dir[39m",
  "[93m   Delete unnecessary files[39m",
]
`;
